File: packages/js/packages/core/src/aad.ts
import { jcs } from "./jcs.js";
import type { VaultContainer, UnlockSlot } from "./types.js";

export function aadObject(container: Pick<
  VaultContainer,
  | "critical_extensions"
  | "crypto"
  | "extensions"
  | "format"
  | "generation"
  | "previous_generation_hash"
  | "unlock_slots"
  | "vault_id"
  | "version"
>): Record<string, unknown> {
  return {
    critical_extensions: container.critical_extensions,
    crypto: { aead: container.crypto.aead, iv: container.crypto.iv },
    extensions: container.extensions,
    format: container.format,
    generation: container.generation,
    previous_generation_hash: container.previous_generation_hash,
    unlock_slots: container.unlock_slots,
    vault_id: container.vault_id,
    version: container.version,
  };
}

export function vaultAad(container: Parameters<typeof aadObject>[0]): Uint8Array {
  return new TextEncoder().encode(jcs(aadObject(container)));
}

export function wrapAad(method: string, slotId: string, vaultId: string): Uint8Array {
  return new TextEncoder().encode(
    jcs({
      method,
      slot_id: slotId,
      vault_id: vaultId,
    }),
  );
}

export function containerWithoutGenerationHash(container: VaultContainer): Record<string, unknown> {
  const { generation_hash: _, ...rest } = container;
  return rest;
}

export function slotFingerprint(slot: UnlockSlot): string {
  return jcs(slot);
}

File: packages/dart/lib/src/aad.dart
import 'dart:typed_data';

import 'jcs.dart';
import 'types.dart';

Map<String, dynamic> aadObject(VaultContainer container) {
  return {
    'critical_extensions':
        container.criticalExtensions.map((e) => e.toJson()).toList(),
    'crypto': {'aead': container.crypto.aead, 'iv': container.crypto.iv},
    'extensions': container.extensions.map((e) => e.toJson()).toList(),
    'format': container.format,
    'generation': container.generation,
    'previous_generation_hash': container.previousGenerationHash,
    'unlock_slots': container.unlockSlots.map((s) => s.toJson()).toList(),
    'vault_id': container.vaultId,
    'version': container.version,
  };
}

Uint8List vaultAad(VaultContainer container) => utf8Encode(jcs(aadObject(container)));

Uint8List wrapAad(String method, String slotId, String vaultId) {
  return utf8Encode(
    jcs({
      'method': method,
      'slot_id': slotId,
      'vault_id': vaultId,
    }),
  );
}

Map<String, dynamic> containerWithoutGenerationHash(VaultContainer container) {
  final json = container.toJson();
  json.remove('generation_hash');
  return json;
}

String slotFingerprint(UnlockSlot slot) => jcs(slot.toJson());
